### ISSUE TITLE
Wrong types when pattern matching `int` with `http.HTTPStatus`

### ISSUE THREAD

**Issue by injust** [python/mypy#18866]
**Bug Report**

When using structural pattern matching to compare an `int` vs an `http.HTTPStatus` instance (and vice versa), mypy gets the type wrong or thinks the code is unreachable.

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.13&flags=strict%2Cwarn-unreachable&gist=843b35aa16cac4c7357b94902cc9057b

```python
from http import HTTPStatus
from typing import reveal_type

code = 200

match code:
    case HTTPStatus():
        pass
    case HTTPStatus.OK:
        reveal_type(code)
    case _:
        pass


status = HTTPStatus.OK

match status:
    case 200:
        reveal_type(status)
    case _:
        pass
```

**Expected Behavior**

The revealed types should be `int` and `Literal[HTTPStatus.OK]`.

**Actual Behavior**

```
main.py:10: note: Revealed type is "Literal[http.HTTPStatus.OK]"
main.py:19: note: Revealed type is "Never"
Success: no issues found in 1 source file
```

**Your Environment**

- Mypy version used: 1.15.0
- Mypy command-line flags: --strict --warn-unreachable
- Python version used: 3.13

**Comment by A5rocks** [comment]
This is kind of https://github.com/python/mypy/issues/17317 but with match statements I think?

**Comment by sterliakov** [comment]
@A5rocks why? #17317 is about accessing enum fields on instances,

```python
from enum import Enum, auto

class Glass(Enum):
    HALF_FULL = auto()
    HALF_EMPTY = auto()

    def print(self) -> None:
        reveal_type(self.HALF_FULL)  # N: Revealed type is "enum.auto"
        reveal_type(Glass.HALF_FULL)  # N: Revealed type is "Literal[__main__.Glass.HALF_FULL]?"
        
        if self == self.HALF_FULL: pass  # E: Non-overlapping equality check (left operand type: "Glass", right operand type: "auto")  [comparison-overlap]
        if self == Glass.HALF_FULL: pass
```

do these accesses have some intimate relationship with enums `match`ing?

**Comment by cdce8p** [comment]
This seems to have been resolved. Tested with `v1.17.1` on Python 3.13.
```py
note: Revealed type is "Literal[http.HTTPStatus.OK]"
note: Revealed type is "Literal[200]"
```
